pkg/controllers/updaterun/initialization_integration_test.go

```diff
@@ -0,0 +1,740 @@
+			Expect(k8sClient.Create(ctx, updateRun)).To(Succeed())
+
+			By("Validating the clusterStagedUpdateRun stats")
+			Eventually(func() error {
+				if err := k8sClient.Get(ctx, updateRunNamespacedName, updateRun); err != nil {
+					return err
+				}
+
+				want := placementv1alpha1.StagedUpdateRunStatus{
+					PolicySnapshotIndexUsed:      policySnapshot.Name,
+					PolicyObservedClusterCount:   numberOfClustersAnnotation,
+					ApplyStrategy:                crp.Spec.Strategy.ApplyStrategy,
+					StagedUpdateStrategySnapshot: &updateStrategy.Spec,
+					StagesStatus: []placementv1alpha1.StageUpdatingStatus{
+						{
+							StageName: "stage1",
+							Clusters: []placementv1alpha1.ClusterUpdatingStatus{
+								{ClusterName: "cluster-9", ClusterResourceOverrideSnapshots: []string{clusterResourceOverride.Name}},
+								{ClusterName: "cluster-7", ClusterResourceOverrideSnapshots: []string{clusterResourceOverride.Name}},
+								{ClusterName: "cluster-5", ClusterResourceOverrideSnapshots: []string{clusterResourceOverride.Name}},
+								{ClusterName: "cluster-3", ClusterResourceOverrideSnapshots: []string{clusterResourceOverride.Name}},
+								{ClusterName: "cluster-1", ClusterResourceOverrideSnapshots: []string{clusterResourceOverride.Name}},
+							},
+							AfterStageTaskStatus: []placementv1alpha1.AfterStageTaskStatus{
+								{Type: placementv1alpha1.AfterStageTaskTypeTimedWait},
+							},
+						},
+						{
+							StageName: "stage2",
+							Clusters: []placementv1alpha1.ClusterUpdatingStatus{
+								{ClusterName: "cluster-0"},
+								{ClusterName: "cluster-2"},
+								{ClusterName: "cluster-4"},
+								{ClusterName: "cluster-6"},
+								{ClusterName: "cluster-8"},
+							},
+							AfterStageTaskStatus: []placementv1alpha1.AfterStageTaskStatus{
+								{
+									Type:                placementv1alpha1.AfterStageTaskTypeApproval,
+									ApprovalRequestName: updateRun.Name + "-stage2",
+								},
+							},
+						},
+					},
+					DeletionStageStatus: &placementv1alpha1.StageUpdatingStatus{
+						StageName: "kubernetes-fleet.io/deleteStage",
+						Clusters: []placementv1alpha1.ClusterUpdatingStatus{
+							{ClusterName: "unscheduled-cluster-0"},
+							{ClusterName: "unscheduled-cluster-1"},
+							{ClusterName: "unscheduled-cluster-2"},
+						},
+					},
+					Conditions: []metav1.Condition{
+						// initialization should succeed!
+						getTrueCondition(updateRun, string(placementv1alpha1.StagedUpdateRunConditionInitialized)),
+					},
+				}
+
+				if diff := cmp.Diff(want, updateRun.Status, cmpOptions...); diff != "" {
+					return fmt.Errorf("status mismatch: (-want +got):\n%s", diff)
+				}
+
+				return nil
+			}, timeout, interval).Should(Succeed(), "failed to validate the clusterStagedUpdateRun in the status")
+		})
+	})
+})
+
+func validateFailedInitCondition(ctx context.Context, updateRun *placementv1alpha1.ClusterStagedUpdateRun, message string) {
```

Comment:
no need to add this in this PR but I think it will be good to also verify the type of error (user error, unexpected, or api error) in this function.

Reply:
This one is related to above discussion. The initialization errors are not returned by the reconciler but just written to the initialization condition. The controller returns nil for these errors to avoid retrying. We can only check the error by checking the condition.